### PR TITLE
Add a gradient checkpoint feature

### PR DESCRIPTION
**Pull request by pass-lin**
[Gradient checkpoint](https://papers.cool/arxiv/1604.06174) is a widely used technique to reduce memory consumption. 
Now we are adapting it for Keras. To make minimal modifications to existing models, we add a parameter `enable_gradient_checkpoint` to the layer, which is set to `False` by default. By simply changing this parameter, we can enable gradient checkpointing. However, for specific implementations depending on different backends, the following points need to be considered:
      In the Torch backend, you should ensure that there are no dropout layers or normalization layers (such as BN, LN, GN, etc.) with inconsistent forward and backward behaviors in the layer of the function you're starting.
      In the TensorFlow backend, you can only enable this setting in eager mode.
      In the JAX backend, you should ensure that there are no strings or other non-differentiable JAX vaild types in the inputs of your function.Such as str

### CODE REVIEW

**Comment by pass-lin** [comment]
@gbaned 
```
All checks passed!
Would reformat: keras/src/layers/layer.py
Would reformat: keras/src/ops/operation.py
2 files would be reformatted, 868 files already formatted
Error: Process completed with exit code 1.
```
Can you teach me how to solve this error?

**Comment by fchollet** [comment]
Thanks for the PR! @divyashreepathihalli has a currently outstanding proposal for implementing gradient checkpointing in Keras across backends. I will you two figure out what to do.

**Comment by pass-lin** [comment]
> Thanks for the PR! @divyashreepathihalli has a currently outstanding proposal for implementing gradient checkpointing in Keras across backends. I will you two figure out what to do.

Thank you. The absence of the gradient checkpoints function in Keras is very troublesome for training LLMs. I hope this new feature can be adapted as soon as possible.

**Comment by divyashreepathihalli** [comment]
@pass-lin thank you for the PR, we appreciate your effort to bring rematerialization support to Keras. However, we are working on adding  this feature and we are trying to add more fine grained control for enabling rematerialization with a mode parameter - more details here - https://docs.google.com/document/d/199s5kaT7fdqDJ5ryJ15aJJH8QIiLvPBYPpb3ZJgPEsE/edit?tab=t.0#heading=h.lleqmh1k4q6g
Once the initial design is worked out, I can open a contribution issue . If you’re interested, you can take up the issue at that point. I’ll make sure to keep you posted as things progress. Closing this PR for now. Thanks again!
